### PR TITLE
Add Chinese zodiac icons (cont)

### LINKED ISSUE #1431: Chinese Zodiac: Switch to Icon template, use new Animal icons
We now have custom animal icons that we can use with the Icon template to really improve this Instant Answer!

The image assets can be found here: http://cl.ly/2Y403l2i253X

------
IA Page:  http://duck.co/ia/view/chinese_zodiac

### PR DESCRIPTION
**Pull request by marianosimone**
Continuing the work started by @Sloff in #1432 (which fixes #1431)

https://duck.co/ia/view/chinese_zodiac

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by marianosimone** [comment]
I know @abeyang asked to remove the dark ring, but that seems to be automatically added by `c-icon__img-wrap--sm` (which makes it consistent with the Zodiac IA)

@moollaza I added a css to include the `wood` color. Any news on getting it added internally instead?

File: lib/DDG/Goodie/ChineseZodiac.pm [hunk 1]
```diff
@@ -11,19 +11,29 @@ use utf8;
...
-    'ox' => { en => 'Ox', zh => '牛' },
-    'tiger' => { en => 'Tiger', zh => '虎' }
+name 'Chinese Zodiac';
+description 'Return the Chinese zodiac animal for a given year';
+primary_example_queries 'chinese zodiac for 1969';
+secondary_example_queries '2004 chinese zodiac animal', 'what was the chinese zodiac animal in 1992', 'what will the chinese zodiac animal be for 2056', 'last year\'s chinese zodiac';
+category 'dates';
+topics 'special_interest';
+code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/ChineseZodiac.pm';
+attribution github => ['http://github.com/wilkox', 'wilkox'],
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: We should remove this metadata - it's all been moved to the platform.

Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @GuiltyDolphin fixed!

File: lib/DDG/Goodie/ChineseZodiac.pm [hunk 2]
```diff
@@ -11,19 +11,19 @@ use utf8;
...
-    'horse' => { en => 'Horse', zh => '马' },
-    'sheep' => { en => 'Goat', zh => '羊' },
-    'monkey' => { en => 'Monkey', zh => '猴' },
-    'fowl' => { en => 'Rooster', zh => '鸡' },
-    'dog' => { en => 'Dog', zh => '狗' },
-    'pig' => { en => 'Pig', zh => '猪' },
-    'rat' => { en => 'Rat', zh => '鼠' },
-    'ox' => { en => 'Ox', zh => '牛' },
-    'tiger' => { en => 'Tiger', zh => '虎' }
+my %animal_to_language_and_path_and_class = (
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: How about `animal_to_attribute`? We usually don't specify every key :wink:

File: t/ChineseZodiac.t [hunk 3]
```diff
@@ -10,36 +10,47 @@ use utf8;
...
+    my $path = $animal_to_language_and_path_and_class{$animal}{'path'};
+    my $class = $animal_to_language_and_path_and_class{$animal}{'class'};
 
     return test_zci("$character ($english)", structured_answer => {
         data => {
             title => "$character ($english)",
-            subtitle => $statement
+            subtitle => $statement,
+            image => $path,
+            url => 'https://en.wikipedia.org/wiki/'.$english.'_(zodiac)'
```
Comment (GuiltyDolphin) [inline t/ChineseZodiac.t]: `"https://en.wikipedia.org/wiki/${english}_(zodiac)"`

File: t/ChineseZodiac.t [hunk 4]
```diff
@@ -10,36 +10,47 @@ use utf8;
...
+            group => "icon",
             item => 0,
-            moreAt => 1
+            moreAt => 1,
+            variants => {
+                iconTitle => 'large',
+                iconImage => 'large'
+            },
+            elClass => {
+                iconImage => $class.' circle'
```
Comment (GuiltyDolphin) [inline t/ChineseZodiac.t]: `"$class circle"`

File: lib/DDG/Goodie/ChineseZodiac.pm [hunk 5]
```diff
@@ -70,26 +70,37 @@ handle remainder => sub {
...
-    my $character = $animal_to_language{$animal}{'zh'};
+    my $english = $animal_to_language_and_path_and_class{$animal}{'en'};
+    my $character = $animal_to_language_and_path_and_class{$animal}{'zh'};
+    my $path = $animal_to_language_and_path_and_class{$animal}{'path'};
+    my $class = $animal_to_language_and_path_and_class{$animal}{'class'};
 
     my $statement = $year_start->strftime("%b %d, %Y") . " – " . $year_end->strftime("%b %d, %Y");
 
-    return format_answer($character, $english, $statement);
+    return format_answer($character, $english, $statement, $path, $class);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: Why don't we just pass in the `$animal` and let `format_answer` grab the attributes? Doesn't look like we do anything else with them here.

**Comment by marianosimone** [comment]
@GuiltyDolphin cleanups applied

File: lib/DDG/Goodie/ChineseZodiac.pm [hunk 6]
```diff
@@ -11,19 +11,19 @@ use utf8;
...
+    'snake' => { en => 'Snake', zh => '蛇' , path => "share/goodie/chinese_zodiac/snake.png", class => "bg-clr--red"},
+    'horse' => { en => 'Horse', zh => '马' , path => "share/goodie/chinese_zodiac/horse.png", class => "bg-clr--red"},
+    'sheep' => { en => 'Goat', zh => '羊' , path => "share/goodie/chinese_zodiac/goat.png", class => "bg-clr--green"},
+    'monkey' => { en => 'Monkey', zh => '猴' , path => "share/goodie/chinese_zodiac/monkey.png", class => "bg-clr--grey"},
+    'fowl' => { en => 'Rooster', zh => '鸡' , path => "share/goodie/chinese_zodiac/rooster.png", class => "bg-clr--grey"},
+    'dog' => { en => 'Dog', zh => '狗' , path => "share/goodie/chinese_zodiac/dog.png", class => "bg-clr--green"},
+    'pig' => { en => 'Pig', zh => '猪' , path => "share/goodie/chinese_zodiac/pig.png", class => "bg-clr--blue-light"},
+    'rat' => { en => 'Rat', zh => '鼠' , path => "share/goodie/chinese_zodiac/rat.png", class => "bg-clr--blue-light"},
+    'ox' => { en => 'Ox', zh => '牛' , path => "share/goodie/chinese_zodiac/ox.png", class => "bg-clr--green"},
+    'tiger' => { en => 'Tiger', zh => '虎' , path => "share/goodie/chinese_zodiac/tiger.png", class => "bg-clr--wood"}
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: Just use the basename here (e.g., `tiger.png`); we can then use `share()->child($path)` to get the images.

File: t/ChineseZodiac.t [hunk 7]
```diff
@@ -10,36 +10,47 @@ use utf8;
...
+    'dragon' => { en => 'Dragon', zh => '龙' , path => "share/goodie/chinese_zodiac/dragon.png", class => "bg-clr--green"},
+    'snake' => { en => 'Snake', zh => '蛇' , path => "share/goodie/chinese_zodiac/snake.png", class => "bg-clr--red"},
+    'horse' => { en => 'Horse', zh => '马' , path => "share/goodie/chinese_zodiac/horse.png", class => "bg-clr--red"},
+    'sheep' => { en => 'Goat', zh => '羊' , path => "share/goodie/chinese_zodiac/goat.png", class => "bg-clr--green"},
+    'monkey' => { en => 'Monkey', zh => '猴' , path => "share/goodie/chinese_zodiac/monkey.png", class => "bg-clr--grey"},
+    'fowl' => { en => 'Rooster', zh => '鸡' , path => "share/goodie/chinese_zodiac/rooster.png", class => "bg-clr--grey"},
+    'dog' => { en => 'Dog', zh => '狗' , path => "share/goodie/chinese_zodiac/dog.png", class => "bg-clr--green"},
+    'pig' => { en => 'Pig', zh => '猪' , path => "share/goodie/chinese_zodiac/pig.png", class => "bg-clr--blue-light"},
+    'rat' => { en => 'Rat', zh => '鼠' , path => "share/goodie/chinese_zodiac/rat.png", class => "bg-clr--blue-light"},
```
Comment (GuiltyDolphin) [inline t/ChineseZodiac.t]: We shouldn't be duplicating data here - just pass in the necessary info in the tests (we only need to test a sub set).

File: lib/DDG/Goodie/ChineseZodiac.pm (hunk 6 above)
Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @GuiltyDolphin: I couldn't get `share()->child($path)` to work, and couldn't find any use of that in the current codebase:

`Can't locate object method "child" via package "Path::Class::Dir"`

Do you have any working example?

File: t/ChineseZodiac.t (hunk 7 above)
Comment (marianosimone) [inline t/ChineseZodiac.t]: I agree about not duplicating data, but it seems like everything in here IS being tested

File: lib/DDG/Goodie/ChineseZodiac.pm (hunk 6 above)
Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @GuiltyDolphin I couldn't get `share()->child($path)` to work, nor I could find any example of its use. I get: `Can't locate object method "child" via package "Path::Class::Dir"`

Do you have any pointers to working usage?

File: t/ChineseZodiac.t (hunk 7 above)
Comment (marianosimone) [inline t/ChineseZodiac.t]: I agree about not duplicating data, but it seems like we ARE testing all those properties

Comment (marianosimone) [inline t/ChineseZodiac.t]: test

File: lib/DDG/Goodie/ChineseZodiac.pm (hunk 6 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: Hmm, I thought it was a `Path::Tiny`, anyway use `share->file($path)` instead.

File: t/ChineseZodiac.t (hunk 7 above)
Comment (GuiltyDolphin) [inline t/ChineseZodiac.t]: @marianosimone Hmm, I guess considering this is the current approach you should be fine.

File: lib/DDG/Goodie/ChineseZodiac.pm (hunk 6 above)
Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: That doesn't return the path either. but a hash. Is anyone using this anywhere? Even the zodiac IA uses this way

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @marianosimone Read `perldoc Path::Class::Dir` - we provide `share` which will give a correct path to the local share directory - paths can go a bit weird at times.

And yes - this is Perl, all 'objects' are hashes.

File: lib/DDG/Goodie/ChineseZodiac.pm [hunk 8]
```diff
@@ -69,27 +69,38 @@ handle remainder => sub {
...
 };
 
 sub format_answer {
-    my ($character, $english, $statement) = @_;
+    my ($statement, $animal) = @_;
+    my $attributes = $animal_attributes{$animal};
+
+    my $english = $attributes->{'en'};
+    my $character = $attributes->{'zh'};
+    my $path = share()->file($attributes->{'icon'})->relative->stringify;
```
Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @GuiltyDolphin what about this?

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @marianosimone That kind of thing - why `relative` though? I'd have thought `absolute` would be more appropriate (unless it doesn't work ;) )

Comment (marianosimone) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @GuiltyDolphin no, absolute doesn't work

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ChineseZodiac.pm]: @marianosimone Fair do-s.

Comment (moollaza) [inline lib/DDG/Goodie/ChineseZodiac.pm]: This won't work in production -- the paths are versioned there. We don't have a great for this fix yet but there are two options.

Grab the current repo version in the Perl, or set the image via JS + the `DDG.get_asset_path()` function

**Comment by marianosimone** [comment]
As pointed out by @moollaza, I'm now getting the repo version in Perl (as there's no JS involved) to get the right path

**Comment by marianosimone** [comment]
@GuiltyDolphin @moollaza ping? Is there anything else needed here?

**Comment by GuiltyDolphin** [comment]
@marianosimone Looks pretty good to me.

@moollaza Can we get this on Beta? Should be mergeable if there's no issues.

**Comment by alohaas** [comment]
@GuiltyDolphin @marianosimone  deploying to beta....

**Comment by moollaza** [comment]
@marianosimone Thanks! This LGTM :+1:

Live on Moollaza: https://moollaza.duckduckgo.com/?q=chinese+zodiac+january+1961&ia=answer

**Pull request merged**